docs/scripts/events.py

```diff
@@ -0,0 +1,142 @@
+
+    all_params = [
+        list(inspect.signature(hook.__init__).parameters.values())[1:]
+        for hook in hooks
+    ]
+
+    # slightly overengineered, but this was fun to write.  ¯\_(ツ)_/¯
+    imports = set()
+    types = set()
+    for params in all_params:
+        for param in params:
+            try:
+                mod = inspect.getmodule(param.annotation).__name__
+                if mod == "typing":
+                    # this is ugly, but can be removed once we are on Python 3.9+ only
+                    imports.add(inspect.getmodule(param.annotation.__args__[0]).__name__)
+                    types.add(param.annotation._name)
+                else:
+                    imports.add(mod)
+            except AttributeError:
+                raise ValueError(f"Missing type annotation: {params}")
+    imports.discard("builtins")
+    if types:
+        print(f"from typing import {', '.join(sorted(types))}")
+    print("from mitmproxy import ctx")
+    for imp in sorted(imports):
+        print(f"import {imp}")
+    print()
+
+    first = True
+    for hook, params in zip(hooks, all_params):
+        if first:
+            first = False
+        else:
+            print()
+        if hook.name in known:
+            raise RuntimeError(f"Already documented: {hook}")
+        known.add(hook.name)
+        doc = inspect.getdoc(hook)
+        print(f"def {hook.name}({', '.join(str(p) for p in params)}):")
+        print(textwrap.indent(f'"""\n{doc}\n"""', "    "))
+        if params:
+            print(f'    ctx.log(f"{hook.name}: {" ".join("{" + p.name + "=}" for p in params)}")')
+        else:
+            print(f'    ctx.log("{hook.name}")')
+    print("```")
+
+
+category(
+    "Lifecycle",
+    [
+        addonmanager.LoadEvent,
+        events.RunningEvent,
+        events.ConfigureEvent,
+        events.DoneEvent,
+    ]
+)
+
+category(
+    "Connection",
+    [
+        server_hooks.ClientConnectedHook,
+        server_hooks.ClientDisconnectedHook,
+        server_hooks.ServerConnectHook,
+        server_hooks.ServerConnectedHook,
+        server_hooks.ServerDisconnectedHook,
+    ]
+)
+
+category(
+    "HTTP",
+    [
+        http.HttpRequestHeadersHook,
+        http.HttpRequestHook,
+        http.HttpResponseHeadersHook,
+        http.HttpResponseHook,
+        http.HttpErrorHook,
+        http.HttpConnectHook,
+    ]
+)
+
+category(
+    "TCP",
+    [
+        tcp.TcpStartHook,
+        tcp.TcpMessageHook,
+        tcp.TcpEndHook,
+        tcp.TcpErrorHook,
+    ]
+)
+
+category(
+    "TLS",
+    [
+        tls.TlsClienthelloHook,
+        tls.TlsStartHook,
+    ]
+)
+
+category(
+    "WebSocket",
+    [
+        websocket.WebsocketStartHook,
+        websocket.WebsocketMessageHook,
+        websocket.WebsocketEndHook,
+        websocket.WebsocketErrorHook,
+    ]
+)
+
+category(
+    "Advanced Lifecycle",
+    [
+        layer.NextLayerHook,
+        events.UpdateEvent,
+        log.AddLogEvent,
+    ]
+)
```

Comment:
Are these only specific events we want to document, or could we auto-generate this, e.g., something like https://stackoverflow.com/a/1796247 ?

Reply:
These are all¹ events, we assert that two lines later. I do like to list/group them manually, this way we can easily get them in a useful order.

¹) all events that were imported at runtime (notice the `import mitmproxy.addons.next_layer  # noqa` at the top). The magic happens [here](https://github.com/mitmproxy/mitmproxy/blob/1158f4e972bd5606bd3d3a38d996c1a8faaf66f9/mitmproxy/events.py#L38). I think this should be good enough. We could conceivably walk the mitmproxy directory for all modules and import them all, but in practice we pretty much achieve full coverage by importing the NextLayer addon.